src/parser/ast/Node.h

```diff
@@ -340,7 +353,7 @@ struct ASTScopeContext : public gc {
     ASTNodeType m_nodeType : 12;
     ASTScopeContextNameInfoVector m_names;
     AtomicStringVector m_usingNames;
-    AtomicStringTightVector m_parameters;
+    ParameterList m_parameters;
```

Comment:
I think this cause memory leak.
because ~ASTScopeContext is never called.
Could you save information as {AtomicString + extra info without saving node}?

Reply:
Firstly, is there any way to check whether it causes memory leak or not?

Secondly, that was a bit hard nut to crack how to achieve to get default arguments. The only problem is (currently since the array/object pattern is not implemented yet) the function expressions, since the bytecode generation of these nodes depend on the current codeBlock's `m_fecounter`. That was the reason why we should push a new scopeContext __before__ we start to parse the function arguments. And all above that escargot has lazy function instantiation, which means the created `DefaultArgumentNode`'s (which was created at the first parsing state) bytecode sequence should be generated only when the function is truly called. But until this point you should keep the created `Node` in alive, since you do not know whether is needed or not, to generate it's bytecode later.

These were the difficulties during the implementation of this feature, do you have any idea/approach to handle this problem better?

Reply:
Yes. you can check memory leak by tool.
There are two allocators in escargot. so we should check those allocators.
you can check builtin allocator(malloc, free) leak by using valgrind
and you can check bdwgc leak on exiting program by this code

```
    GC_gcollect();
    GC_disable();
    GC_enumerate_reachable_objects_inner(
        [](void* obj, size_t bytes, void* cd) {
            size_t size;
            int kind = GC_get_kind_and_size(obj, &size);
            ASSERT(size == bytes);
            void* ptr = GC_USR_PTR_FROM_BASE(obj);
            printf("@@@ kind %d pointer %p\n", (int)kind, ptr);
#if !defined(NDEBUG)
            GC_print_backtrace(ptr);
#endif
        },
        nullptr);
    GC_enable();
```

Reply:
I don't know about default parameter spec completely. 
but I think always {scanning & lazy parse + generating byte code} is better.

Reply:
In my opinion (its just a suggestion), how about expanding the parsing range for each function? Currently, lazy parsing for a function begins its parsing at the start of the function body, not including the parameter list. Instead, when we starts the parsing from the parameter list, we don't have to hold parameter Nodes. Of course, we may need to consider `Arrow Function` differently, but its worth trying.

Reply:
Also, I recommend you to use our inner-defined GC-managed data structures (e.g. `Vector`, `TightVector`) instead of standard library container to avoid any memory leak.

Reply:
I've already had an approach to reparse the function arguments, but it didn't work for the first time due to the scope context mismatch. However, this problem is solved by one part of this PR, so probably it's worth to try it again.